lib/logstash/outputs/elastic_app_search.rb

```diff
@@ -64,13 +65,25 @@ def format_batch(events)
         doc["id"] = event.sprintf(@document_id)
       end
       doc.delete("@version")
-      doc
+      resolved_engine = event.sprintf(@engine)
```

Comment:
Should we warn/error if `sprintf` syntax is used, and the resolved engine is unchanged?

Reply:
Thanks @robbavey for the suggestion. I think that we should warn the user of the missed resolution and remove those documents from the index phase, else they generate an error in AppSearch API call. About this, which is also linked to the handling of `401` and `404` errors, do you think that we should:
-  simply log the warn/error and proceed with the next document?
- log the error and put in DLQ?

Reply:
This might be a good candidate for the DLQ, with a fallback to logging and dropping if DLQ is not available. It could even be a configuration option to block or drop if engine not available? 

Whatever we choose here, we should be up front, and document our intentions clearly

Reply:
Actually if the user set a bad engine the plugin fails at registration time. In the same spirit we could think to have "block" behavior by default, where the block means exit the pipeline with error.

Then add a configuration option that could chose between `:block` or `:drop`, if the user enables `:drop` then it logs a warn, if the DLQ is enabled it also put the offending events in the DLQ.

Reply:
After a discussion with other team members, some doubts to crash a pipeline or event drop has arose, it's simply a bad behavior facing the user with a crash or a drop.

A bad `engine` name could raise from two actions:
- Logstash is not able to resolve (aka `sprintf`) the engine, like in `engine => "%{[not_existing_field]}" with an event that doesn't have any field named `not_existing_field` and this simply resolve to an empty string
- Logstash resolves the engine to a value that's not a valid/configured name in AppSearch application

In the first case is simply a check, in the second the plugin receives an error when issue the http request to the remote AppSearch endpoint.
In both cases we should log a warn, and not drop the event or crash the pipeline, simply the event is not acked in the queue, so the ingestion stops, and the user has ability to fix it.
To help the user with this new behavior we could improve the documentation, providing a sample pipeline example, where there is a last filter that checks for not empty resolution of the engine, and put a default value. This could be a suggestion of a good practice, for example:
```ruby
input {
	stdin {
		codec => json
	}
}

filter {
	if ![engine_name] {
		mutate {
			add_field => {"engine_name" => "default"}
		}
	}
}

output {
	appsearch {
		engine => "%{[engine_name]}"
	}
}
```

Introduce DLQ in this plugin could make bigger the scope of this PR and so in that case I would create a follow up PR for that.

@robbavey what do you think?

Reply:
I think that is a good plan. The most important thing is to be clear and deliberate in what we do, and document our decision clearly - I like the idea of documenting a default. We could potentially allow for a default engine if it doesn't resolve, but the documentation is a good start.

I agree on the DLQ - let's not delay getting this functionality out there on that